Support mangle.nth_identifer to customize base54 behavior

dmichon-msft opened the pull request:
Adds a new option to `options.mangle` and `options.mangle.properties` called `nth_identifier` that allows the caller to completely replace the implementation of `base54`, e.g. to provide identifiers in a particular order, ban certain identifiers, etc.

It is required to be a function with one of the following two interfaces:
```ts
/**
 * An identifier mangler for which the output is invariant with respect to the source code.
 */
export interface SimpleIdentifierMangler {
    /**
     * Obtains the nth most favored (usually shortest) identifier to rename a variable to.
     * The mangler will increment n and retry until the return value is not in use in scope, and is not a reserved word.
     * This function is expected to be stable; Evaluating get(n) === get(n) should always return true.
     * @param n The ordinal of the identifier.
     */
    get(n: number): string;
}

/**
 * An identifier mangler that leverages character frequency analysis to determine identifier precedence.
 */
export interface WeightedIdentifierMangler extends SimpleIdentifierMangler {
    /**
     * Modifies the internal weighting of the input characters by the specified delta.
     * Will be invoked on the entire printed AST, and then deduct mangleable identifiers.
     * @param chars The characters to modify the weighting of.
     * @param delta The numeric weight to add to the characters.
     */
    consider(chars: string, delta: number): number;
    /**
     * Resets character weights.
     */
    reset(): void;
    /**
     * Sorts identifiers by character frequency, in preparation for calls to get(n).
     */
    sort(): void;
}
```

Fixes #719 
Fixes #649

fabiosantoscode commented:
This is looking really good, though I would change the fact that the interface relies on a function with properties. It's pretty hacky and won't map well to what swc is up to with the terser port.

Probably a wrapping object, where one of the properties is this nth_identifier function would work better.

Also a couple things are missing:

 - tools/ contains a typescript definition file that needs to be updated with this new option
 - documentation present in README.md needs to be updated as well.

Thanks for setting this in motion!

dmichon-msft commented:
Changed the signature of the option to be an object, added documentation, added to interface file. Also took the opportunity to move all `reset()`/`sort()` calls into `compute_char_frequency` and `expand_names`.

fabiosantoscode commented:
> Also took the opportunity to move all reset()/sort() calls into compute_char_frequency and expand_names.

Great idea!

I'm surprised this was such a simple change. I had assumed it would touch way too much of the code base. Thanks for doing this :)
